Mathlib/CategoryTheory/Limits/FormalCoproducts.lean

```diff
@@ -0,0 +1,426 @@
+
+
+section Coproduct
+
+variable (𝒜 : Type w) (f : 𝒜 → FormalCoproduct.{w} C) (t X : FormalCoproduct.{w} C)
+
+/-- We construct explicitly the data that specify the coproduct of a given family of formal
+coproducts. -/
+def cofan : Cofan f :=
+  Cofan.mk ⟨(i : 𝒜) × (f i).I, fun p ↦ (f p.1).obj p.2⟩
+    fun i ↦ ⟨fun x ↦ ⟨i, x⟩, fun x ↦ 𝟙 ((f i).obj x)⟩
+
+section simp_lemmas
+
+variable {𝒜 f}
+
+theorem cofan_inj (i : 𝒜) : (cofan 𝒜 f).inj i = ⟨fun x ↦ ⟨i, x⟩, fun x ↦ 𝟙 ((f i).obj x)⟩ := rfl
+-- JH: this is probably a bad theorem? At least not good for simp
+
+@[simp] lemma cofan_inj_f_fst (i : 𝒜) (x) : (((cofan 𝒜 f).inj i).f x).1 = i := rfl
+
+@[simp] lemma cofan_inj_f_snd (i : 𝒜) (x) : (((cofan 𝒜 f).inj i).f x).2 = x := rfl
+
+@[simp] lemma cofan_inj_φ (i : 𝒜) (x) : ((cofan 𝒜 f).inj i).φ x = 𝟙 ((f i).obj x) := rfl
+
+end simp_lemmas
+
+/-- The explicit `Equiv` between maps from the constructed coproduct `cofan 𝒜 f` and families of
+maps from each component, which is the universal property of coproducts. -/
+@[simps!] def cofanHomEquiv :
+    ((cofan 𝒜 f).pt ⟶ t) ≃ ((i : 𝒜) → (f i ⟶ t)) where
+  toFun m i := (cofan 𝒜 f).inj i ≫ m
+  invFun s := ⟨fun p ↦ (s p.1).f p.2, fun p ↦ (s p.1).φ p.2⟩
+  left_inv m := hom_ext rfl (fun ⟨i, x⟩ ↦ by simp [cofan_inj])
+  right_inv p := by ext <;> simp
+
+/-- `cofan 𝒜 f` is a coproduct of `f`. -/
+@[simps!] def isColimitCofan : IsColimit (cofan 𝒜 f) :=
+  mkCofanColimit (cofan 𝒜 f) (fun t ↦ (cofanHomEquiv _ _ _).symm t.inj)
+    (fun t i ↦ congrFun ((cofanHomEquiv _ _ _).right_inv t.inj) i)
+    (fun _ _ h ↦ (Equiv.eq_symm_apply _).2 (funext h))
+
+instance : HasCoproducts.{w} (FormalCoproduct.{w} C) :=
+  hasCoproducts_of_colimit_cofans _ (isColimitCofan _)
+
+/-- The arbitrary choice of the coproduct is isomorphic to our constructed coproduct `cofan 𝒜 f`.
+-/
+noncomputable def coproductIsoCofan : ∐ f ≅ (cofan 𝒜 f).pt :=
+  colimit.isoColimitCocone ⟨_, isColimitCofan _ _⟩
+
+variable {𝒜 f} in
+@[reassoc (attr := simp)] lemma ι_comp_coproductIsoCofan (i) :
+    Sigma.ι f i ≫ (coproductIsoCofan 𝒜 f).hom = (cofan 𝒜 f).inj i :=
+  colimit.isoColimitCocone_ι_hom _ _
+
+/-- Each `X : FormalCoproduct.{w} C` is actually itself a coproduct of objects of the original
+category (after coercion using `of C`). This is the function that specifies the family for which `X`
+is a coproduct of. -/
+def toFun (X : FormalCoproduct.{w} C) : X.I → FormalCoproduct.{w} C :=
+  (of C).obj ∘ X.obj
+
+/-- The witness that each `X : FormalCoproduct.{w} C` is itself a coproduct of objects of the
+original category (after coercion using `of C`), specified by `X.toFun`. -/
+def coproductCoconeIsoSelf : (cofan X.I X.toFun).pt ≅ X :=
+  isoOfComponents (Equiv.sigmaPUnit X.I) fun i ↦ Iso.refl (X.obj i.fst)
+
+@[reassoc (attr := simp)]
+lemma inj_comp_coproductCoconeIsoSelf (i : X.I) :
+    (cofan X.I X.toFun).inj i ≫ (coproductCoconeIsoSelf X).hom = ofHom.mk i (𝟙 (X.obj i)) :=
+  hom_ext rfl (fun i => by simp; rfl)
```

Comment:
```suggestion
  hom_ext rfl (fun i => by simp [coproductCoconeIsoSelf])
```
This works if you `@[simps]` `ofHom.mk` as I suggested above.

Reply:
done